drafts/http-cache-heuristic-freshness.md

```diff
@@ -0,0 +1,86 @@
+# HTTP caching gotcha: Heuristic Freshness
+
+I recently ran into an issue where after deployment of an SPA (Single Page Application), a situation would occur where the page looked broken because CSS could not be loaded.
+
+During analysis of the issue I ran into a thing I had never heard of: *Heuristic Freshness*.
+
+## Context
+
+For context, the application is a typical SPA. The compiled application consists of a bunch of files looking something like this (simplified):
+
+```
+├── index.html
+    └── styles.4a3f9848037579025b00.css
+    └── main.31f7dadf6d2b01fc08c7.js
+```
+
+The browser loads `index.html`, which includes various CSS and JS files.
+
+Caching related headers for the CSS and JS looked like this:
+
+```
+Date: Fri, 06 Dec 2019 13:09:03 GMT
+Etag: "31957a05a5df3c3b315b728b40b6e10e"
+Last-Modified: Mon, 02 Dec 2019 14:12:09 GMT
+Expires:	Sat, 07 Dec 2019 03:09:03 GMT
+```
+
+ `index.html` :
+
+```
+Date:	Fri, 06 Dec 2019 13:09:03 GMT
+Etag: "c4238385fe77f826b5584fed1f1f1659"
+Last-Modified: Tue, 03 Dec 2019 10:50:54 GMT
+```
+
+At first sight things looked okay, but then I noticed there weren't any `Cache-Control` or `Expires` on the `index.html` file.
+
+I'm quite well aware what all these caching headers do when they _are_ present, but I wasn't sure what would happen if they are **not** present 🤔.
+
+## Heuristic Freshness
+
+This brings us to `Heuristic Freshness`. The HTTP specification [defines](https://tools.ietf.org/html/rfc7234#section-4.2.2) that, when a server does *not* explicitly specify expiration times, the client (browser) can use heuristics to estimate a plausible expiration time itself.
+
+How exactly this 'plausible' expiration time is determined is left up to the client, but it seems that in practise most browsers use the following algorithm: `(now() - Last-Modified) * 0.10`. This means two things:
+
+1. You have no control over how long your files are cached in the browser.
```

Comment:
I'm not sure what you mean here. My first reaction is that you do have control, you could just start using the headers. I suppose you mean that once the browser has chosen some expiration time heuristically, you have no control over when the next request will be done.

Reply:
I think my issue here is caused by the word "control". You do have control, you just implicitly gave it up to the browser.

Reply:
What I tried to say was:

When you don't have any `Cache-Control` or `Expires` headers, the browser calculates this plausible expiry time itself. At this point, you don't control for how long your assets are cached in the browser.

Of course, at that point you could start adding those headers but even then it can take some time before the cache is expired and your changes are picked up by the browser.

Reply:
What do you think of this list?

1. When you don't have any `Cache-Control` or `Expires` headers, the browser calculates this plausible expiry time itself.
1. Once your assets are cached, there's no way for you to evict them from the cache.
1. The files will be cached longer as time passes after deployment (assuming your `Last-Modified` headers reflect the time of the last deployment).

Reply:
Sounds good, updated! 👍